local_planner/src/local_footstep_planner.cpp

```diff
@@ -174,47 +171,67 @@ void LocalFootstepPlanner::computeFootPlan(
-                               body_plan.block<1, 3>(horizon_length_ - 1, 6) *
-                                   (midstance - (horizon_length_ - 1)) * dt_;
-          body_rpy_midstance = body_plan.block<1, 3>(horizon_length_ - 1, 3);
-          // grf_midstance = grf_plan.block<1, 3>(horizon_length_ - 1, 3 * j);
+          body_plan_stance = body_plan;
         }
 
+        // Compute the hip position during stance
+        std::vector<Eigen::Vector2d> P, R;
+        for (size_t k = i; k < end_of_stance; k++) {
+          quadKD_->worldToNominalHipFKWorldFrame(
+              j, body_plan_stance.row(k).segment(0, 3),
+              body_plan_stance.row(k).segment(3, 3), hip_position_midstance);
+
+          Eigen::Vector2d p;
+          p << hip_position_midstance.x(), hip_position_midstance.y();
+
+          bool duplicate = false;
+          for (size_t l = 0; l < P.size(); l++) {
+            if ((P.at(l) - p).norm() < 1e-3) {
+              duplicate = true;
+              break;
+            }
+          }
+          if (!duplicate) {
+            P.push_back(p);
+          }
+        }
+
+        // Compute the minimum circle center
+        hip_position_midstance = welzlMinimumCircle(P, R);
+
         // Get touchdown information for body state
         body_vel_touchdown = body_plan.block<1, 3>(i, 6);
         ref_body_vel_touchdown = ref_body_plan.block<1, 3>(i, 6);
         body_ang_vel_touchdown = body_plan.block<1, 3>(i, 9);
         ref_body_ang_vel_touchdown = ref_body_plan.block<1, 3>(i, 9);
 
-        // Compute nominal foot positions for kinematic and grf-projection
-        // measures
-        quadKD_->worldToNominalHipFKWorldFrame(
-            j, body_pos_midstance, body_rpy_midstance, hip_position_midstance);
-        grid_map::Position hip_position_grid_map = {hip_position_midstance.x(),
-                                                    hip_position_midstance.y()};
-
-        auto hip_height =
-            hip_position_midstance.z() -
-            terrain_grid_.atPosition(
-                "z",
-                terrain_grid_.getClosestPositionInMap(hip_position_grid_map),
-                grid_map::InterpolationMethods::INTER_NEAREST);
-
-        centrifugal = (hip_height / 9.81) *
+        // Compute dynamic shift
+        double body_height_touchdown = body_plan(i, 2);
+        // Ref: Highly Dynamic Quadruped Locomotion via Whole-Body Impulse
+        // Control and Model Predictive Control (Centrifugal force and capture
+        // point)
+        centrifugal = body_height_touchdown / 9.81 *
                       body_vel_touchdown.cross(ref_body_ang_vel_touchdown);
-        vel_tracking =
-            (hip_height / 9.81) * (body_vel_touchdown - ref_body_vel_touchdown);
+        // Ref: MIT Cheetah 3: Design and Control of a Robust, Dynamic Quadruped
+        // Robot (Capture point)
+        vel_tracking = std::sqrt(body_height_touchdown / 9.81) *
```

Comment:
Didn't we decide that this didn't need a sqrt?

Reply:
I think that's for the centrifugal term.